### PR TITLE
Add AUM loss for imbalanced classification

### LINKED ISSUE #176: implement AUM loss
for it's performance on unbalanced binary classification problem, as per [AUM paper](https://arxiv.org/pdf/2107.01285)

### PR DESCRIPTION
**Pull request by cregouby**
- refactor both loss to be torch loss
- add AUM loss
- add AUM loss vignette
- update README to better reflect specific features
- refactor optimizer to use `torch::optim_ignite` if available. This provide a 30% speed improvement
image

@dfalbel : would `nn_aum_loss()` be candidate to a {torch} P.R. ?

@tdhock : any comment welcome on the [aum_loss](https://github.com/mlverse/tabnet/blob/a75ec864a6541e08155689906df91aaf61016dec/R/loss.R#L32-L106) and [vignette](https://github.com/mlverse/tabnet/blob/a75ec864a6541e08155689906df91aaf61016dec/vignettes/aum_loss.qmd).

Fix #176 and #177

### CODE REVIEW

File: vignettes/aum_loss.qmd [hunk 1]
```diff
@@ -0,0 +1,235 @@
...
+
+With a class_ratio of 18.1, the target variable is seriously imbalanced, making the minority class much harder to model.
+
+## Solutions to improve imbalanced classification models
+
+First, usual solution to such problem is over-sampling of the minority class, and/or down-sampling the majority class in the training data. We won't cover this here.
+
+The second solution is case weighting. As {tidymodels} offers the framework to manage such case weighting, we'll first use it to compare two model families - XGBoost and Tabnet - with that feature.
+
+Last, we would like to also **optimize** the model according to the metric we are looking at. As the metric of choice for imbalanced dataset are `roc_auc()` or `roc_pr()`, we definitively want a **loss function** that is a proxy of those. This loss is available in {tabnet} with the `nn_aum_loss()` from [Optimizing ROC Curves with a Sort-Based Surrogate Loss for Binary Classification and Changepoint Detection (J Hillman, TD Hocking)](https://jmlr.org/papers/v24/21-0751.html).
```
Comment (tdhock) [inline vignettes/aum_loss.qmd:57]: thanks for the citation!

**Comment by tdhock** [comment]
I did not have time to review everything (this is a big PR with lots of lines changed).
I believe AUM loss could go into torch (rather than tabnet), as it functions similiarly to `torch::nn_bce_with_logits_loss`.

**Comment by cregouby** [comment]
Thanks @tdhock for the remark (and sorry to have mixed AUM loss and refactoring...)
